transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,598 @@
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Background task dedicated to manage the QUIC state machine.
+//!
+//! Considering that all QUIC communications happen over a single UDP socket, one needs to
+//! maintain a unique synchronization point that holds the state of all the active connections.
+//!
+//! The endpoint represents this synchronization point. It is maintained in a background task
+//! whose role is to interface with the UDP socket. Communication between the background task and
+//! the rest of the code only happens through channels. See the documentation of the
+//! [`EndpointDriver`] for a thorough description.
+
+use crate::{
+    connection::Connection,
+    tls,
+    transport::{self, Provider},
+};
+
+use futures::{
+    channel::{mpsc, oneshot},
+    prelude::*,
+    ready,
+};
+use std::{
+    collections::HashMap,
+    io,
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    ops::ControlFlow,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+use x509_parser::nom::AsBytes;
+
+/// Represents the configuration for the QUIC endpoint.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The client configuration to pass to `quinn_proto`.
+    client_config: quinn_proto::ClientConfig,
+    /// The server configuration to pass to `quinn_proto`.
+    server_config: Arc<quinn_proto::ServerConfig>,
+    /// The endpoint configuration to pass to `quinn_proto`.
+    endpoint_config: Arc<quinn_proto::EndpointConfig>,
+}
+
+impl Config {
+    /// Creates a new configuration object with default values.
+    pub fn new(keypair: &libp2p_core::identity::Keypair) -> Result<Self, tls::ConfigError> {
+        let mut transport = quinn_proto::TransportConfig::default();
+        transport.max_concurrent_uni_streams(0u32.into()); // Can only panic if value is out of range.
+        transport.datagram_receive_buffer_size(None);
+        transport.keep_alive_interval(Some(Duration::from_millis(10)));
```

Comment:
It would be nice to set max_idle_timeout (default is 10_000ms).

Reply:
That seems very short. For reference, in go-libp2p we use an idle timeout of 30s, and a keep alive interval of 15s. Obviously, rust-libp2p doesn't need to match these values.

Reply:
We have a test in which we try to dial a non-existing peer. Because right now the limit is 10s this test waits for 10s. It would be nice to have a possibility to set this limit at least for tests.

Reply:
Is there any way to set a shorter timeout for the handshake? 5s should be _plenty_ of time to complete the handshake (a QUIC handshake only takes a single RTT, or 2 RTTs if the peer does a Retry).

Reply:
Set the idle timeout to 30s (but configurable by the user) and added an additional (shorter) timeout for the inital handshake. See 2264ab83377758d0644e668caba1ae9048cc384c.

Reply:
Why is the keep alive period 10ms? That would lead to a keep alive packet to be sent every 10ms. Should this be O(10s) instead? In go-libp2p, we use 15s.